syntax = "proto3";

package package;

import "google/protobuf/empty.proto";

service PackageController {
    rpc List(PackageListRequest) returns (stream Package) {}
    rpc Create(Package) returns (Package) {}
    rpc Retrieve(PackageRetrieveRequest) returns (Package) {}
    rpc Update(Package) returns (Package) {}
    rpc Destroy(Package) returns (google.protobuf.Empty) {}
}

message Package {
    int64 id = 1;
    string key = 2;
    enum PackageStatus {
        DELIVERY_PREPARING = 0;
        DELIVERY_STARTED = 1;
        DELIVERY_FINISHED = 2;
        OVERSEA_SHIPMENT_PREPARING = 3;
        OVERSEA_SHIPMENT_STARTED = 4;
        CANCEL_REQUESTED = 5;
        CANCEL_PENDING = 6;
        CANCEL_FINISHED = 7;
    }
    PackageStatus status = 3;
    repeated string related_order_item_ids = 4;
    string customer_name = 5;
    string customer_contact = 6;
    string base_address = 7;
    optional string detail_address = 8;
    string postal_code = 9;
    optional string delivery_note = 10;
    optional string tracking_number = 11;
    string created_at = 12;
    string updated_at = 13;
    optional string deleted_at = 14;
    optional int64 remark_records = 15;
    optional int64 inventories = 16;
    optional int64 tracking_courier = 17;
}

message PackageListRequest {
}

message PackageRetrieveRequest {
    int64 id = 1;
}
